File: Daily/PY/Leetcode1713-得到子序列的最少操作次数.py
from typing import List
from collections import defaultdict
class Solution:
    def minOperations(self, target: List[int], arr: List[int]) -> int:
        mp = defaultdict(int)
        n = len(target)
        for i in range(n):
            mp[target[i]] = i
        t = []
        rhs = 0
        for val in arr:
            if val in mp:
                idx = mp[val]
                if not t or t[-1] < idx:
                    t.append(idx)
                else:
                    l,r = 0,len(t) - 1
                    # 二分下界
                    while l < r:
                        mid = (l + r) // 2
                        if t[mid] >= idx:
                            r = mid
                        else:
                            l = mid + 1
                    t[l] = idx
        return n - len(t)

File: Daily/sbgo/Leetcode1713-得到子序列的最少操作次数.go
package sbgo

func minOperations(target []int, arr []int) int {
	mp := map[int]int{}
	for i := 0; i < len(target); i++ {
		mp[target[i]] = i
	}
	n := len(target)
	d := []int{}
	for _, val := range arr {
		if idx, has := mp[val]; has {
			if len(d) == 0 || d[len(d)-1] < idx {
				d = append(d, idx)
			} else {
				l, r := 0, len(d)-1
				for l < r {
					mid := (l + r) >> 1
					if d[mid] >= idx {
						r = mid
					} else {
						l = mid + 1
					}
				}
				d[l] = idx
			}
		}
	}
	return n - len(d)
}
